File: backends/credhub/src/main/java/org/cloudfoundry/credhub/certificates/CertificateService.java
package org.cloudfoundry.credhub.certificates;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import org.cloudfoundry.credhub.ErrorMessages;
import org.cloudfoundry.credhub.PermissionOperation;
import org.cloudfoundry.credhub.auth.UserContextHolder;
import org.cloudfoundry.credhub.data.CertificateVersionDataService;
import org.cloudfoundry.credhub.domain.CertificateCredentialVersion;
import org.cloudfoundry.credhub.domain.CredentialVersion;
import org.cloudfoundry.credhub.exceptions.EntryNotFoundException;
import org.cloudfoundry.credhub.services.PermissionCheckingService;

@Service
public class CertificateService {

  private final CertificateVersionDataService certificateVersionDataService;
  private final UserContextHolder userContextHolder;
  private final PermissionCheckingService permissionCheckingService;

  @Autowired
  public CertificateService(
    final CertificateVersionDataService certificateVersionDataService,
    final PermissionCheckingService permissionCheckingService,
    final UserContextHolder userContextHolder) {
    super();
    this.certificateVersionDataService = certificateVersionDataService;
    this.permissionCheckingService = permissionCheckingService;
    this.userContextHolder = userContextHolder;
  }

  public CertificateCredentialVersion findByCredentialUuid(final String uuid) {
    final CredentialVersion credentialVersion = this.certificateVersionDataService
      .findByCredentialUUID(uuid);

    if (!(credentialVersion instanceof CertificateCredentialVersion)) {
      throw new EntryNotFoundException(ErrorMessages.Credential.INVALID_ACCESS);
    }
    final CertificateCredentialVersion certificate = (CertificateCredentialVersion) credentialVersion;
    if (!permissionCheckingService.hasPermission(userContextHolder.getUserContext().getActor(), certificate.getName(), PermissionOperation.READ)) {
      throw new EntryNotFoundException(ErrorMessages.Credential.INVALID_ACCESS);
    }
    return certificate;
  }
}

File: backends/credhub/src/main/kotlin/org/cloudfoundry/credhub/certificates/CertificateService.kt
package org.cloudfoundry.credhub.certificates

import org.cloudfoundry.credhub.ErrorMessages
import org.cloudfoundry.credhub.PermissionOperation.READ
import org.cloudfoundry.credhub.auth.UserContextHolder
import org.cloudfoundry.credhub.data.CertificateVersionDataService
import org.cloudfoundry.credhub.domain.CertificateCredentialVersion
import org.cloudfoundry.credhub.exceptions.EntryNotFoundException
import org.cloudfoundry.credhub.services.PermissionCheckingService
import org.springframework.stereotype.Service

@Service
class CertificateService(
    val certificateVersionDataService: CertificateVersionDataService,
    val permissionCheckingService: PermissionCheckingService,
    val userContextHolder: UserContextHolder
) {

    fun findByCredentialUuid(uuid: String): CertificateCredentialVersion {
        val certificate = certificateVersionDataService.findByCredentialUUID(uuid)
            as? CertificateCredentialVersion ?: throw EntryNotFoundException(ErrorMessages.Credential.INVALID_ACCESS)

        val hasPermission = permissionCheckingService.hasPermission(
            userContextHolder.userContext.actor,
            certificate.name,
            READ
        )

        if (!hasPermission) {
            throw EntryNotFoundException(ErrorMessages.Credential.INVALID_ACCESS)
        }

        return certificate
    }
}
